### ISSUE TITLE
Private packages.json file not pulling in dependencies

### ISSUE THREAD

**Issue by alexbowers** [composer/composer#7578]
I have a URL accessible locally at `http://nova-serve.wip/packages.json` that contains the following
```
{"packages":{"laravel\/nova":{"1.0.0":{"name":"laravel\/nova","version":"1.0.0","version_normalized":"1.0.0.0","dist":{"type":"zip","url":"http:\/\/nova-serve.wip\/nova-1.0.0.zip"}},"1.0.2":{"name":"laravel\/nova","version":"1.0.2","version_normalized":"1.0.2.0","dist":{"type":"zip","url":"http:\/\/nova-serve.wip\/nova-1.0.2.zip"}},"1.0.3":{"name":"laravel\/nova","version":"1.0.3","version_normalized":"1.0.3.0","dist":{"type":"zip","url":"http:\/\/nova-serve.wip\/nova-1.0.3.zip"}}}}}
```

As you can see, there are three versions made available linking to a zip file of the package.

I can now require this in a different project successfully, and it downloads and extracts out the package. That all works as expected.

However, the `installed.json` does not take the `composer.json` file from the package itself, but from the `nova-serve.wip` which I understand is used only to send the package and download location.

How do I get it so that the composer.json file from the package itself is used?

Do I have to have an exact replica of the package.json file for each of the releases on the nova-serve packages list?

**Comment by alexbowers** [comment]
Ok, I can confirm that having the full copy of the composer.json file for nova on the packages.json distribution does work.

Is there a way to get it to use this only for package downloads, and to determine the dependencies from the downloaded package?

**Comment by alcohol** [comment]
Please consider using [`composer/satis`](/composer/satis) if you want to host your own repository.

**Comment by alexbowers** [comment]
The reason for not wanting to use Satis was because the URL provided I would like to be dynamic, based on the user that has logged in, so a static-generated file would not be sufficient for this.

Am i misunderstanding part of the satis project?

**Comment by alcohol** [comment]
EH, good luck then.

**Comment by alexbowers** [comment]
Can i just confirm whether it is expected for the installed.json file to use the result from packages.json on the server and not to use the composer.json file from the package after it is downloaded?

**Comment by alexbowers** [comment]
Ah okay.

It is type `composer`.

Is there an alternative type I can use which is just to point to the download URLs per version, and for composer to take the composer.json values from the package it downloads itself?

**Comment by stof** [comment]
No, because composer won't download the package until the installation phase, when it already decided what it will install. Metadata of packages are needed for the resolution phase.

**Comment by stof** [comment]
Well, you could be writing a composer plugin registering a custom repository type and using that one for your repository, but that would mean that your repository is downloading all versions of all packages available in your repository to get their metadata. Not good for performance
